src/data/list/chain.lean

```diff
@@ -251,23 +251,30 @@ theorem chain'.cons' {x} :
 theorem chain'_cons' {x l} : chain' R (x :: l) ↔ (∀ y ∈ head' l, R x y) ∧ chain' R l :=
 ⟨λ h, ⟨h.rel_head', h.tail⟩, λ ⟨h₁, h₂⟩, h₂.cons' h₁⟩
 
-theorem chain'.drop : ∀ (n) {l} (h : chain' R l), chain' R (drop n l)
-| 0       _             h := h
-| _       []            _ := by {rw drop_nil, exact chain'_nil}
-| (n + 1) [a]           _ := by {unfold drop, rw drop_nil, exact chain'_nil}
-| (n + 1) (a :: b :: l) h := chain'.drop n (chain'_cons'.mp h).right
-
-theorem chain'.append : ∀ {l₁ l₂ : list α} (h₁ : chain' R l₁) (h₂ : chain' R l₂)
-  (h : ∀ (x ∈ l₁.last') (y ∈ l₂.head'), R x y),
-  chain' R (l₁ ++ l₂)
-| []            l₂ h₁ h₂ h := h₂
-| [a]           l₂ h₁ h₂ h := h₂.cons' $ h _ rfl
-| (a :: b :: l) l₂ h₁ h₂ h :=
-  begin
-    simp only [last'] at h,
-    have : chain' R (b :: l) := h₁.tail,
-    exact (this.append h₂ h).cons h₁.rel_head
-  end
+theorem chain'_append : ∀ {l₁ l₂ : list α},
+  chain' R (l₁ ++ l₂) ↔ chain' R l₁ ∧ chain' R l₂ ∧ ∀ (x ∈ l₁.last') (y ∈ l₂.head'), R x y
+| [] l := by simp
+| [a] l := by simp [chain'_cons', and_comm]
+| (a :: b :: l₁) l₂ := by rw [cons_append, cons_append, chain'_cons, chain'_cons, ← cons_append,
+  chain'_append, last', and.assoc]
+
+theorem chain'.append (h₁ : chain' R l₁) (h₂ : chain' R l₂)
+  (h : ∀ (x ∈ l₁.last') (y ∈ l₂.head'), R x y) :
+  chain' R (l₁ ++ l₂) :=
+chain'_append.2 ⟨h₁, h₂, h⟩
+
+theorem chain'.left_of_append (h : chain' R (l₁ ++ l₂)) : chain' R l₁ := (chain'_append.1 h).1
+theorem chain'.right_of_append (h : chain' R (l₁ ++ l₂)) : chain' R l₂ := (chain'_append.1 h).2.1
+
+theorem chain'.infix (h : chain' R l) (h' : l₁ <:+: l) : chain' R l₁ :=
+by { rcases h' with ⟨l₂, l₃, rfl⟩, exact h.left_of_append.right_of_append }
+
+theorem chain'.suffix (h : chain' R l) (h' : l₁ <:+ l) : chain' R l₁ := h.infix h'.is_infix
+theorem chain'.prefix (h : chain' R l) (h' : l₁ <+: l) : chain' R l₁ := h.infix h'.is_infix
+
+theorem chain'.drop (h : chain' R l) (n : ℕ) : chain' R (drop n l) := h.suffix (drop_suffix _ _)
```

Comment:
Can you please add `chain'.take`? Thanks.

Reply:
Done.